core/pta/imx/dek_blob.c

```diff
@@ -0,0 +1,144 @@
+
+#include <drivers/caam_extension.h>
+#include <initcall.h>
+#include <kernel/pseudo_ta.h>
+#include <pta_imx_dek_blob.h>
+#include <string.h>
+#include <tee_api_defines.h>
+#include <tee_api_types.h>
+#include <trace.h>
+
+#define PTA_NAME "dek_blob.pta"
+
+/* Blob size padding in bytes */
+#define BLOB_PAD_SIZE 48
+
+/* HAB Blob header values */
+#define HAB_HDR_TAG	 0x81
+#define HAB_HDR_V4	 0x43
+#define HAB_HDR_MODE_CCM 0x66
+#define HAB_HDR_ALG_AES	 0x55
+
+/*
+ * DEK blobs are stored by the HAB in a secret key blob data structure. Notice
+ * that the HAB supports a set of encryption algorithms, but the encrypted boot
+ * protocol expects AES. The key length is a variable; it can be 128-bit,
+ * 192-bit, or 256-bit.
+ * For more info, see NXP application note AN12056
+ */
+struct dek_blob_header {
+	uint8_t tag;	 /* Constant identifying HAB struct: 0x81 */
+	uint8_t len_msb; /* Struct length in 8-bit msb */
+	uint8_t len_lsb; /* Struct length in 8-bit lsb */
+	uint8_t par;	 /* Constant value, HAB version: 0x43 */
+	uint8_t mode;	 /* AES encryption CCM mode: 0x66 */
+	uint8_t alg;	 /* AES encryption alg: 0x55 */
+	uint8_t size;	 /* Unwrapped key value size in bytes */
+	uint8_t flg;	 /* Key flags */
+} __packed;
+
+/*
+ * Generate HAB DEK blob for encrypted boot
+ *
+ * payload	[in] Plain text key to encapsulate.
+ * payload_size [in] Plain text key size. Must be 128, 192 or 256 bits.
+ * blob		[out] DEK blob.
+ * blob_size	[in/out] DEK blob size.
+ */
+static TEE_Result do_generate(const uint8_t *payload, size_t payload_size,
+			      uint8_t *blob, size_t *blob_size)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	struct dek_blob_header header = {};
+	size_t expected_blob_size = 0;
+	size_t dek_size = 0;
+	uint8_t *dek = NULL;
+
+	assert(payload && blob && payload_size && blob_size);
+	assert(payload_size == (128 / 8) || payload_size == (192 / 8) ||
+	       payload_size == (256 / 8));
+
+	/*
+	 * The DEK size is equals to input key size plus the required blob
+	 * padding. The total output size is the DEK size plus its header
+	 */
+	dek_size = payload_size + BLOB_PAD_SIZE;
+	expected_blob_size = sizeof(header) + dek_size;
+
+	/* Check that the output buffer has the required size */
+	if (*blob_size < expected_blob_size) {
+		*blob_size = expected_blob_size;
+		return TEE_ERROR_SHORT_BUFFER;
+	}
+	*blob_size = expected_blob_size;
+
+	/* Allocate DEK output */
+	dek = calloc(1, dek_size);
+	if (!dek)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	/* Generate DEK */
+	res = caam_dek_generate(payload, payload_size, dek, dek_size);
+	if (res)
+		goto out;
+
+	/* Defined header */
+	header.tag = HAB_HDR_TAG;
+	header.len_msb = 0x00;
+	header.len_lsb = expected_blob_size;
+	header.par = HAB_HDR_V4;
+	header.mode = HAB_HDR_MODE_CCM;
+	header.alg = HAB_HDR_ALG_AES;
+	header.size = payload_size;
+	header.flg = 0x00;
+
+	/* Copy header and the DEK */
+	memcpy(blob, &header, sizeof(header));
+	memcpy(blob + sizeof(header), dek, dek_size);
+out:
+	free(dek);
```

Comment:
maybe `memzero_explicit(dek, ...)` in case there are some sensitive info in the stack.

Reply:
That's not needed, `dek` holds the blob, nothing to protect.
The only confidential data is the plain text key but it does not leave the CAAM.